airbyte-server/src/main/java/io/airbyte/server/converters/DatabaseArchiver.java

```diff
@@ -0,0 +1,138 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.server.converters;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.commons.lang.CloseableConsumer;
+import io.airbyte.commons.lang.Exceptions;
+import io.airbyte.commons.yaml.Yamls;
+import io.airbyte.db.Database;
+import io.airbyte.db.ExceptionWrappingDatabase;
+import io.airbyte.validation.json.JsonSchemaValidator;
+import io.airbyte.validation.json.JsonValidationException;
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.jooq.JSONFormat;
+import org.jooq.JSONFormat.RecordFormat;
+import org.jooq.Record;
+import org.jooq.Table;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DatabaseArchiver {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DatabaseArchiver.class);
+  private static final String DB_FOLDER_NAME = "airbyte_db";
+  private static final String DEFAULT_SCHEMA = "public";
+  private static final JSONFormat DB_JSON_FORMAT = new JSONFormat().recordFormat(RecordFormat.OBJECT);
+
+  private final ExceptionWrappingDatabase database;
+  private final Path storageRoot;
+  private final JsonSchemaValidator jsonSchemaValidator;
+
+  public DatabaseArchiver(final Database database, final Path storageRoot, final JsonSchemaValidator jsonSchemaValidator) {
+    this.database = new ExceptionWrappingDatabase(database);
+    this.storageRoot = storageRoot;
+    this.jsonSchemaValidator = jsonSchemaValidator;
+  }
+
+  public DatabaseArchiver(final Database database, final Path storageRoot) {
+    this(database, storageRoot, new JsonSchemaValidator());
+  }
+
+  /**
+   * Serializes each internal Airbyte Database table into a single archive file stored in YAML.
+   */
+  public void writeDatabaseToArchive() throws IOException {
+    List<Table<?>> tables = listTables();
+    if (tables != null) {
+      tables.forEach(table -> Exceptions.toRuntime(() -> writeTableToArchive(table)));
+    }
+    LOGGER.debug("Successful export of airbyte database");
+  }
+
+  private List<Table<?>> listTables() throws IOException {
+    // list tables from public schema only
+    return database.query(context -> context.meta().getSchemas(DEFAULT_SCHEMA).stream()
+        .flatMap(schema -> context.meta(schema).getTables().stream())
+        .collect(Collectors.toList()));
+  }
+
+  private void writeTableToArchive(Table<?> table) throws Exception {
```

Comment:
do you think we could reasonable move all of the db queries inside `DefaultJobPersistence`? i like that right now all the database interaction is in there and that there is only one place to look. It would definitely be a little clunky because the method would be `dump()` and would have to return stream of stream of json `Stream<Stream<JsonNode>>`. the internal stream is all of the records for a given table. what do you think?

Reply:
it can't be moved in DefaultJobPersistence under two methods? 
- one for listing tables
- one for streaming records for a table?
Is it better to combine everything under a stream of stream?

Reply:
we could do that too. i guess i won't know when we would ever want to list the tables without just doing a dump though.

Reply:
Since I would be adding it to `DefaultJobPersistence` I have to expose those same methods in `JobPersistence` too, right?